src/sync.cpp

```diff
@@ -837,13 +837,34 @@ void Sync::addstatecachechildren(uint32_t parent_dbid, idlocalnode_map* tmap, st
         // clear localname to force newnode = true in setnameparent
         l->localname.clear();
 
-        l->init(this, l->type, p, path);
+        // if we already have the shortname from database, use that, otherwise (db is from old code) look it up
+        std::unique_ptr<string> shortname;
+        if (l->slocalname_in_db)
+        {
+            shortname.reset(l->slocalname.release());
+        }
+        else
+        {
+            shortname.reset(new string());
+            client->fsaccess->getsname(path, shortname.get());
+        }
+
+        l->init(this, l->type, p, path, std::move(shortname));
 
         l->parent_dbid = parent_dbid;
         l->size = size;
         l->setfsid(fsid, client->fsidnode);
         l->setnode(node);
 
+        if (!l->slocalname_in_db)
+        {
+            statecacheadd(l);
+            if (insertq.size() > 50000)
+            {
+                cachenodes();  // periodically output updated nodes with shortname updates, so people who restart megasync still make progress towards a fast startup
```

Comment:
Just a comment. Is it safe to save intermediate updates in permanent cache? It's a great idea to save shortnames during the initial scanning for a faster startup upon restarts, but I'm concerned about applying some changes too early (before completing the scanning), close the app, and upon restart finding an inconsistent state. Perhaps intermediates are harmless, but just in case, to be considered :thinking:

Reply:
I thought it was better to do it this way, because for very large syncs it can take a long time.  If the user kills and restarts MEGAsync, it has to start over, unless we do something like this where we update the database as we go.  In the normal case where all the localnodes already have their shortname looked up, it has no effect because there are no changed localnodes as we load.   It's a minor point though so I don't mind removing it, if you prefer?  thanks

Reply:
Not sure about this one. I understand the advantage when MEGAsync is killed. I'm unsure whether saving partial changes could create any inconsistency afterwards. I leave it at your will :)

Reply:
I think it's ok as it's the same mechanism we use for any other update of a LocalNode's database entry, and the loading code for a LocalNode looks to see if the entry has a shortname present or not.  Let's include it in testing for now and see how it goes, thanks!